### PR TITLE
[WIP] Food Synthisizers!

### PR DESCRIPTION
**Pull request by Poojawa**
It's ya sune, back on their bullshit with another complex but hopefully helpful new machine for overnight shifts and food preferences!

CODE BOUNTY:  ~~TGUI WORKS~~
 $20 to getting crewmember icons to be pulled from datacore without lag. It must be done only by people who have opted in to a prescan or a scan on station. There is likely to be an additional consent check that is opt-in as default as well, but that is not UI related. 

I have fixed the food description issue. WorkingAsIntended ™️ 

Basic replicator machine sprite nabbed from Deep Space 13, which is defunct. I made the screen animations, the panel layer, and the adapted canisters that also can be mounted within.

So there's rounds where there is no chef. 

### This machine does not replace the chef in any way, except the __absence__ of a chef. 

Ok? Ok.

This machine ~~when finished~~ synthesizes a generic, printed block of 1 unit Nutriment. It has no allergy properties and will always give an aftertaste of blandness. 

However, the special notion of this machine is the fact that it will replicate any common food or drink item's looks and (If the food item possesses it) their flavoring. But the replication is merely for people allergic to dairy to be able to eat "Ice Cream". There may also be special functions, should one take a looksie and have experience working with wires. (hint hint)

Science can upgrade the machine, Improving scanners improves how fast the machine will print a chosen food item, improving manipulators grants more nutriment per printing, Rank 5 = ~2s printing, and 5 nutriment per food item. So science has a reason to do something helpful for the masses if they so choose. Most chef products *start* at 3 and scale up significantly with recipes and added ingredient quality. 

The synthesizers have "Synthesizer cartridge" fitted within them. They contain a reagent that cannot be produced locally on station and must be ordered from cargo. This choice is so that it isn't an infinite resource to take advantage of. The normal sized canisters have sufficient supply for 50 meals, whereas the smaller one (intended for use on shuttles/Talon/etc) will only merely do 20. replacing them is as easy as slapping a new can into it!

### CODE REVIEW

File: tgui/packages/tgui/interfaces/synthesizer.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import { filter, sortBy } from 'common/collections';
+import { Fragment } from 'inferno';
+import { useBackend, useSharedState } from "../backend";
+import { Box, Button, Flex, Icon, Input, LabeledList, ProgressBar, Section, Dropdown } from "../components";
+import { Window } from "../layouts";
+import { createSearch, toTitleCase } from 'common/string';
+
+const canBeMade = (recipe) => {
+  return true;
+};
```
Comment (ItsSelis) [inline tgui/packages/tgui/interfaces/synthesizer.js:12]:
```suggestion
```

File: tgui/packages/tgui/interfaces/synthesizer.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                  Object
+                    .keys(recipe.requirements)
+                    .map(mat => toTitleCase(mat) + ": " + recipe.requirements[mat])
+                    .join(", ")
+                ) || (
+                  <Box>
+                    No resources required.
+                  </Box>
+                )}
+              </Flex.Item>
```
Comment (ItsSelis) [inline tgui/packages/tgui/interfaces/synthesizer.js:91]:
```suggestion
```

File: tgui/packages/tgui/interfaces/synthesizer.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                      x5
+                    </Button>
+                    <Button
+                      color={recipe.hidden && "red" || null}
+                      disabled={!canBeMade(recipe, materials, 10)}
+                      onClick={() => act("make", { make: recipe.ref, multiplier: 10 })}>
+                      x10
+                    </Button>
+                  </Box>
+                ) || null}
```
Comment (ItsSelis) [inline tgui/packages/tgui/interfaces/synthesizer.js:78]:
```suggestion
```

File: tgui/packages/tgui/interfaces/synthesizer.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            onInput={(e, v) => setSearchText(v)}
+            mb={1} />
+          {recipesToShow.map(recipe => (
+            <Flex justify="space-between" align="center" key={recipe.ref}>
+              <Flex.Item>
+                <Button
+                  color={recipe.hidden && "red" || null}
+                  icon="hammer"
+                  iconSpin={busy === recipe.name}
+                  disabled={!canBeMade(recipe, materials, 1)}
```
Comment (ItsSelis) [inline tgui/packages/tgui/interfaces/synthesizer.js:59]:
```suggestion
```

File: tgui/packages/tgui/interfaces/synthesizer.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
+import { round } from 'common/math';
+import { flow } from 'common/fp';
+import { filter, sortBy } from 'common/collections';
+import { Fragment } from 'inferno';
+import { useBackend, useSharedState } from "../backend";
+import { Box, Button, Flex, Icon, Input, LabeledList, ProgressBar, Section, Dropdown } from "../components";
+import { Window } from "../layouts";
+import { createSearch, toTitleCase } from 'common/string';
```
Comment (ItsSelis) [inline tgui/packages/tgui/interfaces/synthesizer.js:8]:
```suggestion
import { flow } from 'common/fp';
import { filter, sortBy } from 'common/collections';
import { useBackend, useSharedState } from "../backend";
import { Box, Button, Flex, Input, Section, Dropdown } from "../components";
import { Window } from "../layouts";
import { createSearch, toTitleCase } from 'common/string';
```

File: tgui/packages/tgui/interfaces/synthesizer.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+const canBeMade = (recipe) => {
+  return true;
+};
+
+export const synthesizer = (props, context) => {
```
Comment (ItsSelis) [inline tgui/packages/tgui/interfaces/synthesizer.js:14]:
```suggestion
export const Synthesizer = (props, context) => {
```

File: code/modules/food/kitchen/cooking_machines/synthesizer_recipes_vr.dm [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+/*************
+* Categories *
+*************/
+
+/datum/category_group/synthesizer
+
+/datum/category_group/synthesizer/all
+	name = "All"
+	category_item_type = /datum/category_item/synthesizer
+
```
Comment (ItsSelis) [inline code/modules/food/kitchen/cooking_machines/synthesizer_recipes_vr.dm:55]:
```suggestion

/datum/category_group/synthesizer/all/New()
	..()

```

File: code/modules/food/kitchen/cooking_machines/synthesizer_vr.dm [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+
+	circuit = /obj/item/weapon/circuitboard/synthesizer
+	var/datum/wires/synthesizer/wires = null
+
+	//loaded cartridge
+	var/obj/item/weapon/reagent_containers/synth_disp_cartridge/cart
+	var/cart_type = /obj/item/weapon/reagent_containers/synth_disp_cartridge
+
+	//all of our food
+	var/static/datum/category_collection/synthesizer_recipes = list()
```
Comment (ItsSelis) [inline code/modules/food/kitchen/cooking_machines/synthesizer_vr.dm:42]:
```suggestion
	var/static/datum/category_collection/synthesizer_recipes/synthesizer_recipes
```

File: code/modules/food/kitchen/cooking_machines/synthesizer_vr.dm [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+			var/percent = round((C.reagents.total_volume / C.volume) * 100)
+			. += "The installed cartridge has [percent]% remaining."
+
+	return
+
+// TGUI to do.
+/obj/machinery/synthesizer/tgui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "synthesizer", name)
```
Comment (ItsSelis) [inline code/modules/food/kitchen/cooking_machines/synthesizer_vr.dm:84]:
```suggestion
		ui = new(user, src, "Synthesizer", name)
```

File: code/modules/food/kitchen/cooking_machines/synthesizer_vr.dm [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+	return ..()
+
+/obj/machinery/synthesizer/tgui_static_data(mob/user)
+	var/list/data = ..()
+
+	var/list/categories = list()
+	var/list/recipes = list()
+	for(var/datum/category_group/synthesizer/A in synthesizer_recipes.categories)
+		categories += A.name
+		for(var/datum/category_item/synthesizer/F in A)
```
Comment (ItsSelis) [inline code/modules/food/kitchen/cooking_machines/synthesizer_vr.dm:99]:
```suggestion
		for(var/datum/category_item/synthesizer/F in A.items)
```

**Review by ItsSelis — CHANGES_REQUESTED** [review]
The fixes you asked for.

**Comment by ItsSelis** [comment]
Additionally, the file at `tgui/packages/tgui/interfaces/synthesizer.js` should be renamed into Synthesizer.js (just capitalizing).

File: code/modules/food/kitchen/cooking_machines/synthesizer_vr.dm [hunk 11]
```diff
@@ -0,0 +1,539 @@
...
+				//Sanity check.
+				if(!making || !src)
+					return
+				to_chat(world,"We have the supplies for [making.name], onwards to printing!")
+				busy = TRUE
+				update_use_power(USE_POWER_ACTIVE)
+				update_icon() // light up time
+				playsound(src, 'sound/machines/replicator_input_ok.ogg', 100)
+				C.reagents.remove_reagent("synthsoygreen", 5)
+				var/obj/item/weapon/reagent_containers/food/snacks/food_mimic = new making.path(src)
```
Comment (Poojawa) [inline code/modules/food/kitchen/cooking_machines/synthesizer_vr.dm:191]: I'd really like for this food to actually, y'know. copy shit over.

**Comment by Cameron653** [comment]
Doing PR cleanup and the last commit to this was 2 months ago and there are conflicts that need to be resolved.

If work is still being done on this PR, feel free to re-open this PR or open a new PR.

**Pull request closed without merging**